packages/flutter_test/lib/src/binding.dart

```diff
@@ -507,6 +507,17 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
   }
 
   Future<Null> _runTestBody(Future<Null> testBody(), VoidCallback invariantTester) async {
+    // Delay this function by a microtask.
+    //
+    // The `_runTest` functions invokes `_runTestBody` and then invokes
+    // `asyncBarrier()` afterwardes. The `asyncBarrier` verifies that no scope is
+    // open.
+    // We don't want to move the `asyncBarrier()` earlier, because it also
+    // flushes all microtasks.
+    // By delaying by one microtask, the `asyncBarrier` will not yet see the
+    // open scope and also ensure that the all microtasks (including this one)
+    // are flushed.
+    await new Future<Null>.microtask(() {});
```

Comment:
(github ate my comment so I'm going to copy and paste it here, my apologies:)

I think what I'm having trouble with is the question of what the "open scope" that this comment is referring to is. Where do we open a scope which delaying the code here helps with? This microtask doesn't interact with the async barrier, since the async barrier isn't called until later within this method (synchronously) anyway. Indeed, the async barrier can't possibly be what drives the microtasks (as per the last sentence of the comment here) since it isn't invoked until after the microtasks are invoked, since it's called synchronously by this method.

Reply:
The `_runTestBody` function calls `runApp` followed by `pump` which seems to open a scope.
After the call to `pump` it yields (since it's doing an `await`), and the controlflow continues in `_runTest` (line 504 above) where it does a `timeout?.catchError` followed by an `asyncBarrier`.
The `asyncBarrier` checks that no scope is open, which is not the case, since `pump` just opened one.

By delaying the `_runTestBody` function the `_runTest` continues *first* with the `timeout?.catchError` and `asyncBarrier` (now succeeding) before running the `_runTestBody`.

Reply:
This shouldn't be an issue anymore.

Reply:
ah, interesting. sounds like the old code was just wrong too, then. The purpose of the guards is to catch open scopes in the test that weren't fully awaited.